source/agora/consensus/protocol/Nominator.d

```diff
@@ -465,6 +474,51 @@ extern(D):
             log.trace("SCP indicated invalid envelope: {}", scpPrettify(&envelope));
     }
 
+    /***************************************************************************
+
+        Called when a new Block Signature is received from the network.
+
+        Params:
+            block_sig = the SCP envelope
+
+        Returns:
+            true if the SCP protocol accepted this envelope
+
+    ***************************************************************************/
+
+    public void receiveBlockSignature (scope ref const(ValidatorBlockSig) block_sig) @trusted
+    {
+        const cur_height = this.ledger.getBlockHeight();
+        log.trace("Store signature for slot #{} (ledger is at height #{}) for node {}",
+                block_sig.height, cur_height, block_sig.public_key);
+        const Point K = Point(block_sig.public_key[]);
+        this.slot_sigs[block_sig.height][K] = block_sig.signature;
+
+        auto blocks = this.ledger.getBlocksFrom(Height(block_sig.height));
+        if (!blocks.empty)
+        {
+
+            Block block = cast(Block) blocks.front;
```

Comment:
Why the unsafe cast? The other routines should just take the block as const.

Reply:
Do you think I should add dup() to Block and Header to make it possible to create the mutable Block needed?

Reply:
I mean why does it need to be mutable? If you really are modifying it then casting const breaks assumptions & safety.

Reply:
So if I understand correctly I should build a new block within the called function. So pass in a `const block` and add an `out signed_block` as a parameter to return the block with the updated header fields?

Reply:
Actually I suppose I should create a new function in `Block` to create a new `const Block` with only the `signature` and `validators` fields updated in the `BlockHeader`.

Reply:
Pushed an update to remove the cast